File: datastructure/sorting/quicksort.cpp
#include<bits/stdc++.h>
using namespace std;
int partition(vector<int>& nums, int low, int high) {
    int pivot = nums[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
        if (nums[j] <= pivot) {
            i++;
            swap(nums[i], nums[j]);
        }
    }
    swap(nums[i + 1], nums[high]);
    return i + 1;
}

void quicksort(vector<int>& nums, int low, int high) {
    if (low < high) {
        int pi = partition(nums, low, high);
        quicksort(nums, low, pi - 1);
        quicksort(nums, pi + 1, high);
    }
}

int main() {
    int elements;
    cin >> elements;
    vector<int> nums(elements);
    for (int i = 0; i < elements; i++) {
        cin >> nums[i];
    }
    quicksort(nums, 0, elements - 1);
    for (int i = 0; i < nums.size(); i++) {
        cout << nums[i] << " ";
    }
    cout << endl;

    return 0;
}

File: datastructure/sorting/quicksort.py
def partion(nums, low, high):
    pivot = nums[high]
    i = low - 1
    for j in range(low, high):
        if nums[j] <= pivot:
            i += 1
            nums[i], nums[j] = nums[j], nums[i]
    nums[i + 1], nums[high] = nums[high], nums[i + 1]
    return i + 1

def quicksort(nums, low, high):
    if low < high:
        pi = partion(nums, low, high)
        quicksort(nums, low, pi - 1)
        quicksort(nums, pi + 1, high)

nums = list(map(int, input().split()))
quicksort(nums, 0, len(nums) - 1)
print(nums)
